File: Rust/src/body.rs
use crate::utils::vec2::Vector2;

pub struct Body {
  pub position: Vector2,
  pub rotation: f32,  
  pub velocity: Vector2,
  pub angular_velocity: f32,

  pub width: Vector2,
  pub friction: f32,
  pub mass: f32,
  pub inverse_mass: f32,
  pub inertia: f32,
  pub inverse_inertia: f32,

  pub torque: f32,
  pub force: Vector2
}

impl Body {
  pub fn new() -> Self {
    Self {
      position: Vector2::new(0.0, 0.0),
      rotation: 0.0,
      velocity: Vector2::new(0.0, 0.0),
      angular_velocity: 0.0,

      width: Vector2::new(1.0, 1.0),
      friction: 0.2,
      mass: f32::MAX,
      inverse_mass: 0.0,
      inertia: f32::MAX,
      inverse_inertia: 0.0,

      torque: 0.0,
      force: Vector2::new(0.0, 0.0),
    }
  }

  pub fn set(&mut self, width: &Vector2, mass: f32) {
    self.position.set(0.0, 0.0);
    self.rotation = 0.0;
    self.velocity.set(0.0, 0.0);
    self.angular_velocity = 0.0;

    self.width = *width;
    self.friction = 0.0;
    self.mass = mass;

    if mass < f32::MAX {
      self.inverse_mass = 1.0 / mass;

      self.inertia = mass * (
        width.x * width.x + width.y * width.y
      ) / 12.0;

      self.inverse_inertia = 1.0 / self.inertia;
    }

    else {
      self.inverse_mass = 0.0;
      self.inertia = f32::MAX;
      self.inverse_inertia = 0.0;
    }

    self.torque = 0.0;
    self.force.set(0.0, 0.0);
  }

  pub fn add_force(&mut self, force: &Vector2) {
    self.force += *force;
  }
}

File: Zig/src/body.zig
const std = @import("std");
const Vec2 = @import("./root.zig").vec.Vec2;
const expect = @import("std").testing.expect;

pub const Body = struct
{
    position: Vec2 = Vec2 { .x = 0, .y = 0 },
    rotation: f32 = 0,

    velocity: Vec2 = Vec2 { .x = 0, .y = 0 },
    angular_velocity: f32 = 0,

    width: Vec2 = Vec2 { .x = 1, .y = 1 },
    friction: f32 = 0.2,

    force: Vec2 = Vec2 { .x = 0, .y = 0 },
    torque: f32 = 0,

    mass: f32 = std.math.floatMax(f32),
    inv_mass: f32 = 0,

    i: f32 = std.math.floatMax(f32),
    inv_i: f32 = 0,

    pub fn set(this: *Body, width: *const Vec2, mass: f32) void
    {
        this.position.set(0, 0);
        this.rotation = 0;

        this.velocity.set(0, 0);
        this.angular_velocity = 0;

        this.width.assign(width);
        this.friction = 0;

        this.force.set(0, 0);
        this.torque = 0;

        this.mass = mass;

        if (mass < std.math.floatMax(f32))
        {
            this.inv_mass = 1 / mass;
            this.i = mass * (width.x * width.x + width.y * width.y) / 12;
            this.inv_i = 1 / this.i;
        }
        else
        {
            this.inv_mass = 0;
            this.i = std.math.floatMax(f32);
            this.inv_i = 0;
        }
    }

    pub fn addForce(this: *Body, force: *const Vec2) void
    {
        this.force.addAssign(force);
    }
};
